packages/rum-core/src/common/context.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2017-present, Elasticsearch BV
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import Url from './url'
+import { PAGE_LOAD } from './constants'
+import { getPageMetadata, getServerTimingInfo } from './utils'
+
+/**
+ * Both Navigation and Resource timing level 2 exposes these below information
+ *
+ * for CORS requests without Timing-Allow-Origin header, transferSize & encodedBodySize will be 0
+ */
+function getResponseContext(perfTimingEntry) {
+  const {
+    transferSize,
+    encodedBodySize,
+    decodedBodySize,
+    serverTiming
+  } = perfTimingEntry
+
+  const respContext = {
+    transfer_size: transferSize,
+    encoded_body_size: encodedBodySize,
+    decoded_body_size: decodedBodySize
+  }
+  const serverTimingStr = getServerTimingInfo(serverTiming)
+  if (serverTimingStr) {
+    respContext.headers = {
+      'server-timing': serverTimingStr
+    }
+  }
+  return respContext
+}
+
+function getResourceContext(data) {
+  const { entry, url } = data
+  return {
+    http: {
+      url,
+      response: getResponseContext(entry)
+    }
+  }
+}
+
+function getExternalContext(data) {
+  const { url, method, target, response } = data
+  const parsedUrl = new Url(url)
+
+  const { href, port, protocol, hostname, origin } = parsedUrl
+  const isDefaultPort = port === '80' || port === '443'
```

Comment:
this should really be dependent on the protocol/scheme

Reply:
That part is taken care here right? https://github.com/elastic/apm-agent-rum-js/blob/b0d89b4ef23e8d3ece2dcbd30f383a7dcdf3c4e2/packages/rum-core/src/common/url.js#L149 Am i missing something?

Reply:
You're _setting_ the default port there, but (and I know this is pathological) what if the URI explicitly specifies the port, like `https://host:80`? Port 80 is not the default port for https.

Reply:
if there is a explicit mention of port, it wont hit this `else` block right?  we would be using 80 for the port.

Reply:
> if there is a explicit mention of port, it wont hit this else block right? we would be using 80 for the port.

In the code you linked to, yeah. But I'm talking about `function getExternalContext`, which is only looking at the resulting port number (which may be the default, or may have been explicitly specified) without also looking at the scheme, to decide that it's a default port.

In case it's still not clear what I'm on about, add case to `'should add external span context with all fields'` like this:

```js
url = 'https://[::1]:80/'
span = createSpanWithData(url)
```

IIANM, your code will set `destination.service.name` to `https://[::1]`, but it should be `https://[::1]:80` as 80 is not the default port for https.

Reply:
oops, good point. Sorry for the confusion and thanks for explaining. Will fix it.